src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -282,166 +292,195 @@ struct TreeVisitor<'tree> {
+        (self.f_continue)(&args)
+    }
+
+    fn propagate_at(
         &mut self,
         this: &mut TreeVisitor<'_>,
         idx: UniIndex,
-        exclude: Option<UniIndex>,
         rel_pos: AccessRelatedness,
     ) -> Result<(), OutErr> {
-        // 1. apply the propagation function
         let node = this.nodes.get_mut(idx).unwrap();
-        let recurse =
-            (self.f_propagate)(NodeAppArgs { node, perm: this.perms.entry(idx), rel_pos })
-                .map_err(|error_kind| {
-                    (self.err_builder)(ErrHandlerArgs {
-                        error_kind,
-                        conflicting_info: &this.nodes.get(idx).unwrap().debug_info,
-                        accessed_info: &this.nodes.get(self.initial).unwrap().debug_info,
-                    })
-                })?;
-        let node = this.nodes.get(idx).unwrap();
-        // 2. add the children to the stack for future traversal
-        if matches!(recurse, ContinueTraversal::Recurse) {
-            let general_child_rel = rel_pos.for_child();
-            for &child in node.children.iter() {
-                // Some child might be excluded from here and handled separately,
-                // e.g. the initially accessed tag.
-                if Some(child) != exclude {
-                    // We should still ensure that if we don't skip the initially accessed
-                    // it will receive the proper `AccessRelatedness`.
-                    let this_child_rel = if child == self.initial {
-                        AccessRelatedness::This
-                    } else {
-                        general_child_rel
-                    };
-                    self.stack.push((child, this_child_rel));
+        (self.f_propagate)(NodeAppArgs { node, perm: this.perms.entry(idx), rel_pos }).map_err(
+            |error_kind| {
+                (self.err_builder)(ErrHandlerArgs {
+                    error_kind,
+                    conflicting_info: &this.nodes.get(idx).unwrap().debug_info,
+                    accessed_info: &this.nodes.get(self.initial).unwrap().debug_info,
+                })
+            },
+        )
+    }
+
+    fn go_upwards_from_accessed(
+        &mut self,
+        this: &mut TreeVisitor<'_>,
+        accessed_node: UniIndex,
+        push_children_of_accessed: bool,
+    ) -> Result<(), OutErr> {
+        assert!(self.stack.is_empty());
+        // First, handle accessed node. A bunch of things need to
+        // be handled differently here compared to the further parents
+        // of `accesssed_node`.
+        {
+            self.propagate_at(this, accessed_node, AccessRelatedness::This)?;
+            if push_children_of_accessed {
+                let accessed_node = this.nodes.get(accessed_node).unwrap();
+                for &child in accessed_node.children.iter().rev() {
+                    self.stack.push((child, AccessRelatedness::AncestorAccess, false));
```

Comment:
Please either use an `enum` instead of just `false` here, or make this something like `/* after_children */ false` or so.

Reply:
This is not resolved yet -- there's no enum and no comment.

I think an enum would be better.

Reply:
Ah whoops, I misidentified this line. I thought you meant the argument to that method, not the transient internal element that is pushed onto the encapsulated stack.

Reply:
I really don't know what to call them, except for `Phase1` and `Phase2`, or maybe `BeforeTheRecursiveCallsIfThisWasWrittenAsARecursiveFunction` and `AfterTheRecursiveCallsIfThisWasWrittenAsARecursiveFunction`.